### PR TITLE
Add warning for random.random 

### PR DESCRIPTION
**Pull request by Eddy114514**
add warning for random.random() that in python 3 its string representation is longer (~18 digit) than python 2 (~14 digit).

as following:
image
image
image

### CODE REVIEW

**Comment by ltratt** [comment]
Please force push a rebase against master before review.

**Comment by Eddy114514** [comment]
> Please force push a rebase against master before review.

@ltratt , resolved, thanks for pointing out.

**Comment by ltratt** [comment]
Please squash.

**Comment by ltratt** [comment]
Hmm, actually, maybe I'm wrong.

Isn't this warning going to be annoying? Even if someone *has* updated their code, they're still going to get the "change your code" warning AFAICS?

**Comment by Eddy114514** [comment]
> Hmm, actually, maybe I'm wrong.
> 
> Isn't this warning going to be annoying? Even if someone _has_ updated their code, they're still going to get the "change your code" warning AFAICS?

@ltratt I think you are right about the warning behavior of my implementation. I think it is generally hard to detect whether user change their code within the warning. 

Since the warning is about length of random.random(), maybe I should put warning inside of `len()` or `__str__` method of random.random (if there is one) such that it is less frequent. But I think it still can't detect whether user change their code or not.

**Comment by ltratt** [comment]
What one would need to do is track the string output by `random.random` and then see if someone does something "bad" on it (e.g. `len`). There are various ways one could do this.

For example, @nanjekyejoannah can fill you in on the `bstate` idea we've long been working towards. A variant of that would work here: one _could_ (say) add a `bool` to PyStr that is `true` if the string was produced by `random.random`. This is pretty heavy-weight, admittedly.

Another way would be to return a "special" string-like object from `random` and intercept calls to it. I'm not sure how hard that might be.

**Comment by Eddy114514** [comment]
> What one would need to do is track the string output by `random.random` and then see if someone does something "bad" on it (e.g. `len`). There are various ways one could do this.
> 
> For example, @nanjekyejoannah can fill you in on the `bstate` idea we've long been working towards. A variant of that would work here: one _could_ (say) add a `bool` to PyStr that is `true` if the string was produced by `random.random`. This is pretty heavy-weight, admittedly.
> 
> Another way would be to return a "special" string-like object from `random` and intercept calls to it. I'm not sure how hard that might be.

@ltratt Thank you for your suggestions on improving my implementation, I will try to work on either ways and figure out which one is more suitable. Thanks again!

**Comment by Eddy114514** [comment]
@ltratt 
After discussed and get help from @nanjekyejoannah about bstate yesterday, I choose to return a special float-like object (subclass of float object) from` random.random`. And since `random.random `return float object, I move check from `len` of string object to `__str__` of float-like object (I called it pg_float). 

I avoid putting a new filed inside of float object since change size of struct might cause some chain reaction.

**Comment by ltratt** [comment]
This looks interesting but dumb question: doesn't this still mean that everyone who does `len(random.random())` gets the warning i.e. there's no way to silence it?

**Comment by Eddy114514** [comment]
> This looks interesting but dumb question: doesn't this still mean that everyone who does `len(random.random())` gets the warning i.e. there's no way to silence it?

@ltratt it seems that len(random.random()) won't run because float objects doesn't have such method. As picture shown below

**Comment by ltratt** [comment]
Sorry, I meant `len(str(random.random()))`.

**Comment by Eddy114514** [comment]
@ltratt I think since we are intended to warn user about the length change of `str(random.random())`, so user should get warning by directly calling `len()` on it. And when user change their code, for code like `if (len(str(random.random())) == 14 (length of random.random() in python2)`, their could be 2 way:

- `if(len(str(random.random())[:14]) == 14`
- `if(len(str(random.random())) == 18 `

We could avoid warning in first case that we also store the extra information about whether generated from random in string-object and check in `len()` method that whether it is from `random` and with length of 18. But for second case, unless we also store the information inside of integer return by `len()`, we might can't avoid false positive.

We could maybe also maintain a global dictionary with key be location of warning (code line number, etc) and value be a bool of whether user change their code when user run pygrate3 on their code that we assume user's change is correct and mark bool at that location to be true and avoid throw warning to user in next check. But if we assume user's change is correct then 1 run of pygrate3 is enough such that we don't even need to worry about silence problem.

**Comment by nanjekyejoannah** [comment]
@Eddy114514  lets first find a justification for duplicate C code on floats to avoid duplication.

File: Modules/_randommodule.c [hunk 1]
```diff
@@ -177,7 +177,14 @@ _random_Random_random_impl(RandomObject *self)
 /*[clinic end generated code: output=117ff99ee53d755c input=afb2a59cbbb00349]*/
 {
     uint32_t a=genrand_uint32(self)>>5, b=genrand_uint32(self)>>6;
-    return PyFloat_FromDouble((a*67108864.0+b)*(1.0/9007199254740992.0));
+    double x = (a * 67108864.0 + b) * (1.0 / 9007199254740992.0);
+    if (Py_Py2xWarningFlag){
+        return _PyFloat_FromDoubleWithFlags(x,0x01);
```
Comment (nanjekyejoannah) [inline Modules/_randommodule.c]: small style issue

File: Objects/floatobject.c [hunk 2]
```diff
@@ -385,6 +396,14 @@ float_repr(PyFloatObject *v)
         return PyErr_NoMemory();
     result = _PyUnicode_FromASCII(buf, strlen(buf));
     PyMem_Free(buf);
+    
+    if (Py_Py2xWarningFlag && ((v->ob_pg_flags & 0x01) != 0) && PyErr_WarnEx(
+        PyExc_Py2xWarning,
+        "String repr of random.random() is longer in 3.x, change code accordingly",
```
Comment (nanjekyejoannah) [inline Objects/floatobject.c]: Make fix more specific, @ltratt ideas?

**Comment by ltratt** [comment]
If we want to move this PR forwards I think we need two passing tests:

1. A test that the warning for `len` on the string produced by `random` produces a warning.
2. A test that strings produced in other ways don't produce a warning.

**Comment by Eddy114514** [comment]
> If we want to move this PR forwards I think we need two passing tests:
> 
> 1. A test that the warning for `len` on the string produced by `random` produces a warning.
> 2. A test that strings produced in other ways don't produce a warning.

@ltratt 
Thanks for the idea!
image
I believe my current implementation and updated test address such requirement.
I also encountered some difficulties in squashing commits that because there are some PR passed during the process of this PR, if I squash all commits from random, it will include those that doesn't belong to this PR.

File: Lib/test/test_py2xwarn.py [hunk 3]
```diff
@@ -38,6 +38,14 @@ def test_next(self):
             w.reset()
             self.assertNoWarning(iterator_marks.__next__(), w)
             
+    def test_random(self):
+        expected = "String repr of random.random() is longer in 3.x, change code accordingly"
+        import random
+        with check_py2x_warnings(("", Py2xWarning)) as w:
+            self.assertWarning(len(str(random.random())), w, expected)
+            w.reset()
+            self.assertNoWarning(len(str(0.123456)), w)
```
Comment (nanjekyejoannah) [inline Lib/test/test_py2xwarn.py:47]: Test LGTM

**Comment by ltratt** [comment]
Please squash.

**Comment by Eddy114514** [comment]
> Please squash.

@ltratt squashed

**Comment by Eddy114514** [comment]
@ltratt the test logs suggest that I need to remove or configure some test cases for changed struct of float and unicode object, will do.

**Comment by ltratt** [comment]
What's the status of this one?

**Comment by Eddy114514** [comment]
> What's the status of this one?

@ltratt Sorry for the delay, I was taking  ans preparing for finals this and last week. Today I will take my last final, will work on it ASAP

**Comment by Eddy114514** [comment]
@ltratt I fixed the issue about test_sys and test on my local machine, it should be able to pass the status checks. Thanks

File: Include/cpython/floatobject.h [hunk 4]
```diff
@@ -5,12 +5,17 @@
 typedef struct {
     PyObject_HEAD
     double ob_fval;
+    unsigned char ob_randomflag;
 } PyFloatObject;
 
+#define NOT_FROM_RANDOM 0
```
Comment (ltratt) [inline Include/cpython/floatobject.h]: This looks like a `bool` to me? :) I assume CPython can use `stdbool`? If so, we should just use that.

File: Modules/_randommodule.c [hunk 5]
```diff
@@ -177,7 +177,14 @@ _random_Random_random_impl(RandomObject *self)
 /*[clinic end generated code: output=117ff99ee53d755c input=afb2a59cbbb00349]*/
 {
     uint32_t a=genrand_uint32(self)>>5, b=genrand_uint32(self)>>6;
-    return PyFloat_FromDouble((a*67108864.0+b)*(1.0/9007199254740992.0));
+    double x = (a * 67108864.0 + b) * (1.0 / 9007199254740992.0);
+    if (Py_Py2xWarningFlag){
+        return _PyFloat_FromDoubleWithFlags(x,FROM_RANDOM);   
+    }
```
Comment (ltratt) [inline Modules/_randommodule.c]: Formatting snafu.

File: Objects/floatobject.c [hunk 6]
```diff
@@ -376,14 +387,18 @@ float_repr(PyFloatObject *v)
...
     buf = PyOS_double_to_string(PyFloat_AS_DOUBLE(v),
                                 'r', 0,
                                 Py_DTSF_ADD_DOT_0,
                                 NULL);
     if (!buf)
         return PyErr_NoMemory();
-    result = _PyUnicode_FromASCII(buf, strlen(buf));
+    if (Py_Py2xWarningFlag){
+        result = _PyUnicode_FromASCII_withRandomFlag(buf, strlen(buf), v->ob_randomflag);
+    }
```
Comment (ltratt) [inline Objects/floatobject.c]: Same.

File: Include/cpython/floatobject.h (hunk 4 above)
Comment (Eddy114514) [reply Include/cpython/floatobject.h]: @ltratt  I suppose bool and unsigned char have same size and using unsigned char can allow us to prepare for future extension (e.g if there are >2 status) 

Style issue are fixed and squashed.

Comment (ltratt) [reply Include/cpython/floatobject.h]: I really think this would be clearer as a `bool`, because there is no likelihood that we have more than 2 states here.

Comment (Eddy114514) [reply Include/cpython/floatobject.h]: > I really think this would be clearer as a `bool`, because there is no likelihood that we have more than 2 states here.

@ltratt Thanks for pointing out, I have switched to bool for float struct, and tested on my local machine and squashed changes.

File: Modules/_randommodule.c [hunk 7]
```diff
@@ -177,7 +177,13 @@ _random_Random_random_impl(RandomObject *self)
 /*[clinic end generated code: output=117ff99ee53d755c input=afb2a59cbbb00349]*/
 {
     uint32_t a=genrand_uint32(self)>>5, b=genrand_uint32(self)>>6;
-    return PyFloat_FromDouble((a*67108864.0+b)*(1.0/9007199254740992.0));
+    double x = (a * 67108864.0 + b) * (1.0 / 9007199254740992.0);
+    if (Py_Py2xWarningFlag)
+        return _PyFloat_FromDoubleWithFlags(x, FROM_RANDOM);
+    else{
```
Comment (ltratt) [inline Modules/_randommodule.c]: I think we should either use brackets on both true/false branches, or on neither. And to keep the code similar to its surrounds `{` should be preceded by a space.

File: Objects/unicodeobject.c [hunk 8]
```diff
@@ -11605,7 +11616,15 @@ unicode_join(PyObject *self, PyObject *iterable)
 
 static Py_ssize_t
 unicode_length(PyObject *self)
-{
+{   
+    if(Py_Py2xWarningFlag && PyUnicode_Check(self)){
```
Comment (ltratt) [inline Objects/unicodeobject.c]: Any reason not to use `&&` for the clause below too?

File: Objects/floatobject.c [hunk 9]
```diff
@@ -376,14 +387,17 @@ float_repr(PyFloatObject *v)
...
     buf = PyOS_double_to_string(PyFloat_AS_DOUBLE(v),
                                 'r', 0,
                                 Py_DTSF_ADD_DOT_0,
                                 NULL);
     if (!buf)
         return PyErr_NoMemory();
-    result = _PyUnicode_FromASCII(buf, strlen(buf));
+    if (Py_Py2xWarningFlag)
+        result = _PyUnicode_FromASCII_withRandomFlag(buf, strlen(buf), v->ob_randomflag);
+    else{
```
Comment (ltratt) [inline Objects/floatobject.c]: Same syntax comment.

File: Include/cpython/floatobject.h (hunk 4 above)
Comment (Eddy114514) [reply Include/cpython/floatobject.h]: @ltratt thanks for your instruction, i have fixed the syntax problem, also I have to revert to unsigned char because I can't import stdbool in public header in cpython, as the following picture suggest. 
image

Comment (ltratt) [reply Include/cpython/floatobject.h]: I might be missing something obvious but this still seems to be `unsigned char`?

Comment (ltratt) [reply Include/cpython/floatobject.h]: @Eddy114514 I think we're still waiting on this comment.

**Comment by Eddy114514** [comment]
@ltratt 

Hi, I think I might misused the replying system, so my previous comments might not have been visible to you.
Regarding the change: I had to revert to using `unsigned char` because importing `<stdbool.h>` in a public header is not allowed in CPython. As shown in the screenshot below, including `stdbool.h` triggers a build error.image
My previous comments are as following:
image

**Comment by ltratt** [comment]
Typically one includes these headers in `.c` files, not `.h` files, to avoid this problem.

**Comment by Eddy114514** [comment]
@ltratt Thank you for the suggestion.
Since `PyFloatObject` is defined in `floatobject.h` , including `<stdbool.h>` in `floatobject.c ` wouldn’t allow us to change the struct field type to bool.
Maybe I could keep `ob_randomflag `as an unsigned char in the public struct, and include `<stdbool.h>` in` floatobject.c` to convert it to a local bool for better readability where it is used. Please let me know what you think. Thank you.

**Comment by nanjekyejoannah** [comment]
Lets explore some more on how to support bool @Eddy114514 it might just require some gimics than normal

**Comment by Eddy114514** [comment]
@ltratt @nanjekyejoannah After discovering several ways, I figure out that I could use `_Bool` in struct in `flaotobject.h `such that it avoid `include <stdbool.h>` and avoid the massive changes for moving struct in ``flaotobject.c`. I have squashed and tested on my local machine. Thank for your help

File: Include/cpython/floatobject.h (hunk 4 above)
Comment (ltratt) [reply Include/cpython/floatobject.h]: We should get rid of this `#define` now we use `_Bool`. That might mean the bool flag should be renamed to `_ob_is_random` or similar (i.e. to make the flag itself clear).

Comment (Eddy114514) [reply Include/cpython/floatobject.h]: > We should get rid of this `#define` now we use `_Bool`. That might mean the bool flag should be renamed to `_ob_is_random` or similar (i.e. to make the flag itself clear).

@ltratt Thanks for your suggestions, I have refactored the name globally for both float and Unicode object, and remove the define to use 0 and 1 for false and true.

**Comment by Eddy114514** [comment]
@ltratt  I have squashed the changes
